content/sips/sip-366.md

```diff
@@ -16,36 +16,277 @@ proposal: >-
+4. Update the `escrowedCollateralByAccount` to reflect the collateral amount declared in the intent for the identified account and pool.
+5. Update the `intentNoncesByAccount` to include the new intent nonce associated with the account.
+6. Update the `intentByNonce` to represent the newly declared intent using the specific nonce.
+7. Emit a `DelegateCollateralIntentDeclared` event.
+
+#### Processing Intents
+
+Add the following function to **process** an outstanding intent to delegate a new amount of collateral (refer to the NatSpec annotations for further implementation guidance):
+
+```solidity
+/**
+ * @notice Attempt to process an outstanding intent to delegate a
+ * new amount of collateral
+ * @dev The {msg.sender} can be any address
+ * @param accountId The ID of the account for which the outstanding
+ * intent to delegate a new amount of collateral is being processed
+ * @param poolId The ID of the pool for which the intent of the account
+ * to delegate a new amount of collateral is being processed
+ * @param collateralType The address of the collateral type that the
+ * account has an outstanding intent to delegate a new amount of
+ */
+function processIntentToDelegateCollateral(
+    uint128 accountId,
+    uint128 poolId,
+    address collateralType
+) public;
+```
+
+The `processIntentToDelegateCollateral` function should, at minimum, satisfy the following requirements:
+
+1. Allow anyone to call it without requiring permissions, enabling the processing of any open intent.
+2. Verify the existence of the specified intent for the given account, pool, and type of collateral.
+3. Ensure the intent's submission timestamp is greater than the current timestamp plus the applicable delay.
+4. Ensure the intent's submission timestamp does not surpass the current timestamp plus both the delay and window period.
+5. Check that processing the intent won't cause the collateralization ratio to fall below the required issuance ratio.
+6. Update the `escrowedCollateralByAccount` to mirror the amount of collateral in process for the designated account and pool.
+7. Delete the intent nonce from `intentNoncesByAccount` for the involved account.
+8. Delete the intent from `intentByNonce` using the specific nonce.
+9. Update the amount of collateral delegated to the pool according to the intent.
+10. Emit a `DelegateCollateralIntentProcessed` event upon successful processing.
+11. Execute steps 2 to 11 for every intent ready for processing for the mentioned account, pool, and collateral type.
+
+> The order by which the intents are processed should not matter so long as the `escrowedCollateralByAccount` mapping field is properly maintained.
+
+##### Edge Cases
+
+If an attempt is made to process an intent but constraints such as issuance ratio restrictions and debt fluctuations prevent the full amount of collateral from being processed, then the maximum permissible amount of collateral should be determined and processed as allowed.
```

Comment:
Is it possible to get to a situation where a user has more set to be undelegated than what he has delegated? Possibly due to partial liquidation? In that case the maximum amount of collateral is undelegated.

Reply:
yes, but the spec here should cover that as well. I updated